alf/algorithms/dqn_algorithm.py

```diff
@@ -0,0 +1,252 @@
+        else:
+            alpha = torch.exp(self._log_alpha[0]).detach()
+        # p(a|s) = exp(Q(s,a)/alpha) / Z;
+        logits = q_values / alpha
+        discrete_action_dist = td.Categorical(logits=logits)
+        if eps_greedy_sampling:
+            discrete_action = dist_utils.epsilon_greedy_sample(
+                discrete_action_dist,
+                epsilon_greedy,
+                # NOTE: This is the only departure from SAC:
+                True)
+        else:
+            discrete_action = dist_utils.sample_action_distribution(
+                discrete_action_dist)
+
+        action_dist = discrete_action_dist
+        action = discrete_action
+
+        if (self._reproduce_locomotion and rollout
+                and not self._training_started):
+            # This uniform sampling seems important because for a squashed Gaussian,
+            # even with a large scale, a random policy is not nearly uniform.
+            action = alf.nest.map_structure(
+                lambda spec: spec.sample(outer_dims=observation.shape[:1]),
+                self._action_spec)
+
+        return action_dist, action, q_values, new_state
+
+    # Copied and modified from sac_algorithm (discrete actions).
+    def rollout_step(self, inputs: TimeStep, state: DqnState):
+        """``rollout_step()`` basically predicts actions like what is done by
+        ``predict_step()``. Additionally, if states are to be stored a in replay
+        buffer, then this function also call ``_critic_networks`` and
+        ``_target_critic_networks`` to maintain their states.
+        """
+        action_dist, action, _, action_state = self._predict_action(
+            inputs.observation,
+            state=state.action,
+            # NOTE: This is the only departure from SAC.
+            epsilon_greedy=self._rollout_epsilon_greedy(),
+            eps_greedy_sampling=True,
+            rollout=True)
+
+        if self.need_full_rollout_state():
+            _, critics_state = self._compute_critics(
+                self._critic_networks, inputs.observation, action,
+                state.critic.critics)
+            _, target_critics_state = self._compute_critics(
+                self._target_critic_networks, inputs.observation, action,
+                state.critic.target_critics)
+            critic_state = DqnCriticState(
+                critics=critics_state, target_critics=target_critics_state)
+            actor_state = ()
+        else:
+            actor_state = state.actor
+            critic_state = state.critic
+
+        new_state = DqnState(
+            action=action_state, actor=actor_state, critic=critic_state)
+        return AlgStep(
+            output=action,
+            state=new_state,
+            info=DqnInfo(action=action, action_distribution=action_dist))
+
+    # Copied and modified from sac_algorithm (discrete actions).
+    def _critic_train_step(self, inputs: TimeStep, state: DqnCriticState,
```

Comment:
I suggest that you just define a one-hot action distribution (argmax entry has 1) for DQN and directly reuse SAC's `_critic_train_step`. You can do this in your defined `_predict_action` (first return result `action_dist`).